src/elfxx.c

```diff
@@ -445,6 +445,91 @@ elf_w (find_section) (struct elf_image *ei, const char* secname)
   return 0;
 }
 
+
+static char *
+elf_w (add_hex_byte) (char *str, uint8_t byte)
+{
+  const char hex[] = "0123456789abcdef";
+
+  *str++ = hex[byte >> 4];
+  *str++ = hex[byte & 0xf];
+  *str = 0;
+
+  return str;
+}
+
+
+static int
+elf_w (find_build_id_path) (const struct elf_image *ei, char *path, unsigned path_len)
+{
+  const Elf_W (Ehdr) *ehdr = ei->image;
+  const Elf_W (Shdr) *shdr;
+  unsigned i;
+
+  if (!elf_w (valid_object) (ei))
+    return -1;
+
+  shdr = elf_w (section_table) (ei);
+  if (!shdr)
+    return -1;
+
+  for (i = 0; i < ehdr->e_shnum; ++i, shdr = (const Elf_W (Shdr) *) (((const uint8_t *) shdr) + ehdr->e_shentsize))
+    {
+      const uint8_t *notes;
+      const uint8_t *notes_end;
+
+      /* The build-id is in a note section */
+      if (shdr->sh_type != SHT_NOTE)
+        continue;
+
+      notes = ((const uint8_t *) ei->image) + shdr->sh_offset;
+      notes_end = notes + shdr->sh_size;
+
+      while(notes < notes_end)
+        {
+          const char prefix[] = "/usr/lib/debug/.build-id/";
+
+          /* See "man 5 elf" for notes about alignment in Nhdr */
+          const Elf_W(Nhdr *) nhdr = (const ElfW(Nhdr *)) notes;
+          const ElfW(Word) namesz = nhdr->n_namesz;
+          const ElfW(Word) descsz = nhdr->n_descsz;
+          const ElfW(Word) nameasz = UNW_ALIGN(namesz, 4); /* Aligned size */
+          const char *name = (const char *) (nhdr + 1);
+          const uint8_t *desc = (const uint8_t *) name + nameasz;
+          unsigned j;
+
+          notes += sizeof(*nhdr) + nameasz + UNW_ALIGN(descsz, 4);
+
+          if ((namesz != sizeof(ELF_NOTE_GNU)) ||  /* Spec says must be "GNU" with a NULL */
```

Comment:
ELF_NOTE_GNU is not defined non-GNU platforms. All of this code needs to be wrapped in `#if defined(ELF_NOTE_GNU)`.

Reply:
I wonder when ELF_NOTE_GNU and NT_GNU_BUILD_ID were added to glibc? A bit of digging suggests that NT_GNU_BUILD_ID didn't appear until 2.7 (ELF_NOTE_GNU is earlier). Is there a minimum version of glibc header files that libunwind supports? I can fix this another way if necessary.

Reply:
It should be enough to just `#ifdef` directly on those macros since they could be supplied by implementations other than GNU for compatibility.